plan/plans/row_stack.go

```diff
@@ -0,0 +1,203 @@
+	r, b, err := p.Src.Filter(ctx, expr)
+	if !b || err != nil {
+		return p, false, errors.Trace(err)
+	}
+
+	p.Src = r
+	return p, true, nil
+}
+
+// Next implements the plan.Plan Next interface.
+func (p *RowStackFromPlan) Next(ctx context.Context) (*plan.Row, error) {
+	row, err := p.Src.Next(ctx)
+	if row == nil || err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	row.FromData = row.Data
+
+	updateRowStack(ctx, nil, row.Data)
+
+	return row, nil
+}
+
+// Close implements the plan.Plan Close interface.
+func (p *RowStackFromPlan) Close() error {
+	return p.Src.Close()
+}
+
+// A dummy type to avoid naming collision in context.
+type rowStackKeyType int
+
+// String defines a Stringer function for debugging and pretty printing.
+func (k rowStackKeyType) String() string {
+	return "row stack"
+}
+
+// rowStackKey is used to retrive outer table references for sub query.
+const rowStackKey rowStackKeyType = 0
+
+// RowStack saves the current and outer row references.
+// For every select, we will push a RowStack to a stack for inner sub query use,
+// so the top RowStack is always for current select.
+// e.g, select c1 from t1 where c2 = (select c1 from t2 where t2.c1 = t1.c2 limit 1),
+// the "select c1 from t1" is the outer query for the sub query in where phase, we will
+// first push a RowStack to the stack saving the row data for "select c1 from t1", then
+// push the second RowStack to the stack saving the row data for "select c1 from t2".
+// We will push a RowStack after the from phase and pop it before the final phase.
+// So we can guarantee that there is at least one RowStack certainly.
+type RowStack struct {
+	items []*rowStackItem
+}
+
+type rowStackItem struct {
+	// OutDataFields is the output record data with select list.
+	OutData       []interface{}
+	OutDataFields []*field.ResultField
+	// FromData is the first origin record data, generated by From.
+	FromData       []interface{}
+	FromDataFields []*field.ResultField
+}
+
+func getRowStack(ctx context.Context) *RowStack {
+	v := ctx.Value(rowStackKey)
+	if v == nil {
+		return nil
+	}
+	// must be RowStack
+	t := v.(*RowStack)
+	return t
+}
+
+func pushRowStack(ctx context.Context, outDataFields []*field.ResultField, fromDataFields []*field.ResultField) {
+	s := getRowStack(ctx)
+	if s == nil {
+		s = &RowStack{
+			items: make([]*rowStackItem, 0, 1),
+		}
+	}
+
+	s.items = append(s.items, &rowStackItem{
+		OutDataFields:  outDataFields,
+		FromDataFields: fromDataFields,
+	})
+
+	ctx.SetValue(rowStackKey, s)
+}
+
+func updateRowStack(ctx context.Context, outData []interface{}, fromData []interface{}) error {
+	s := getRowStack(ctx)
+	if s == nil {
+		return errors.Errorf("update empty row stack")
+	}
+
+	t := s.items[len(s.items)-1]
+	t.OutData = outData
+	t.FromData = fromData
+	return nil
+}
+
+func popRowStack(ctx context.Context) error {
+	s := getRowStack(ctx)
+
+	if s == nil || len(s.items) == 0 {
+		return errors.Errorf("pop empty row stack")
```

Comment:
no, it will not. 
We push RowStack before Next in final plan, then pop it. If we have empty result set, the top item in RowStack just has no data, an it is ok.